include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,499 @@
+#define _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_INCLUDED_
+
+#include <nbl/builtin/hlsl/impl/emulated_float64_t_impl.hlsl>
+namespace nbl
+{
+namespace hlsl
+{
+    template<bool FastMath = true, bool FlushDenormToZero = true>
+    struct emulated_float64_t
+    {
+        using storage_t = uint64_t;
+
+        storage_t data;
+
+        // constructors
+        /*static emulated_float64_t create(uint16_t val)
+        {
+            return emulated_float64_t(bit_cast<uint64_t>(float64_t(val)));
+        }*/
+
+        static emulated_float64_t<FastMath, FlushDenormToZero> create(int32_t val)
+        {
+            return emulated_float64_t<FastMath, FlushDenormToZero>(bit_cast<uint64_t>(float64_t(val)));
+        }
+
+        static emulated_float64_t<FastMath, FlushDenormToZero> create(int64_t val)
+        {
+            return emulated_float64_t<FastMath, FlushDenormToZero>(bit_cast<uint64_t>(float64_t(val)));
+        }
+
+        static emulated_float64_t<FastMath, FlushDenormToZero> create(uint32_t val)
+        {
+            return emulated_float64_t<FastMath, FlushDenormToZero>(bit_cast<uint64_t>(float64_t(val)));
+        }
+
+        static emulated_float64_t<FastMath, FlushDenormToZero> create(uint64_t val)
+        {
+            return emulated_float64_t<FastMath, FlushDenormToZero>(bit_cast<uint64_t>(float64_t(val)));
+        }
+
+        static emulated_float64_t<FastMath, FlushDenormToZero> create(float64_t val)
+        {
+            return emulated_float64_t<FastMath, FlushDenormToZero>(bit_cast<uint64_t>(val));
+        }
+        
+        // TODO: unresolved external symbol imath_half_to_float_table
+        /*static emulated_float64_t create(float16_t val)
+        {
+            return emulated_float64_t(bit_cast<uint64_t>(float64_t(val)));
+        }*/
+
+        static emulated_float64_t create(float32_t val)
+        {
+            return emulated_float64_t(bit_cast<uint64_t>(float64_t(val)));
+        }
+
+        static emulated_float64_t createPreserveBitPattern(uint64_t val)
+        {
+            return emulated_float64_t(val);
+        }
+
+        // arithmetic operators
+        emulated_float64_t operator+(const emulated_float64_t rhs)
+        {
+            emulated_float64_t retval = createPreserveBitPattern(0u);
+
+            uint64_t mantissa;
+            uint32_t3 mantissaExtended;
+            int biasedExp;
+
+            uint64_t lhsSign = data & ieee754::traits<float64_t>::signMask;
+            uint64_t rhsSign = rhs.data & ieee754::traits<float64_t>::signMask;
+            uint64_t lhsMantissa = ieee754::extractMantissa(data);
+            uint64_t rhsMantissa = ieee754::extractMantissa(rhs.data);
+            int lhsBiasedExp = ieee754::extractBiasedExponent(data);
+            int rhsBiasedExp = ieee754::extractBiasedExponent(rhs.data);
+            
+            int expDiff = lhsBiasedExp - rhsBiasedExp;
+
+            if (lhsSign == rhsSign)

```

Comment:
the amount of if/else is a bit concerning here

Reply:
not anymore